AOC [23.02](https://adventofcode.com/2023/day/2)

Elf plays a game with you. Pulls cubes from bag. Each game has XXX samples from bag.

Game XX: sample1; sample2; . . . sampleN
samples separated by semicolon

Which games possible with a bag of 12 red, 13 green, 14 blue?

To solve: look at each game XX and see if it has fewer than the number of possible cubes

1. read line
1. split at colon (get game N)
1. split game data at semicolons (get sample)
1. split samples at comma - get color data & cube N data


In [1]:
line = 'Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green'
head,_,tail = line.partition(':')
print(head,tail)


Game 1  3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green


In [2]:
import re
findNumbers  = re.compile(r'\d')
findColors = re.compile(r'(red)|(green)|(blue)')


gameNum = re.findall(findNumbers,head)[0]
print(gameNum)

1


In [3]:
samples = tail.split(';')
print(samples)

[' 3 blue, 4 red', ' 1 red, 2 green, 6 blue', ' 2 green']


In [4]:
for sample in samples:
    sampData = sample.split(',')
    print(sampData)

[' 3 blue', ' 4 red']
[' 1 red', ' 2 green', ' 6 blue']
[' 2 green']


In [5]:
re.search(findColors,'2 green')[0]

'green'

In [8]:
maxVals = {'red':12,'green':13,'blue':14}

for sample in samples:
    drawGroup = sample.split(',')
    for draw in drawGroup:
        nCubes = int(re.search(findNumbers,draw)[0])
        color = re.search(findColors,draw)[0]
        if nCubes < maxVals[color]:
            impossible = True
        else:
            impossible = False
    print(drawGroup,impossible)



[' 3 blue', ' 4 red'] True
[' 1 red', ' 2 green', ' 6 blue'] True
[' 2 green'] True


OK, I've got each of the bits sorted. Let's put it together. . . 

I will createa a function for the test sample: as soon as one of the values > maxValues, it will return false -- that way we can stop testing as soon as the sample is unviable.

In [10]:
testData='''Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green'''
import re
findNumbers  = re.compile(r'\d+')
findColors = re.compile(r'(red)|(green)|(blue)')

def testSample(samples):
    for sample in samples.split(';'):
        drawGroup = sample.split(',')
        for draw in drawGroup:
            nCubes = int(re.search(findNumbers,draw)[0])
            color = re.search(findColors,draw)[0]
            if nCubes > maxVals[color]:
                return False
    return True

maxVals = {'red':12,'green':13,'blue':14}
tally = 0

for line in testData.split('\n'):
    head,_,tail = line.partition(':')
    gameNum = re.search(findNumbers,head)[0]
    samples = tail.split(';')
    if testSample(tail):
        tally += int(gameNum)

print(tally)      
    
    



8


test works, run it on the data

In [11]:
with open ('2302input.txt') as f_in:
    data = f_in.read().split('\n')

import re
findNumbers  = re.compile(r'\d+')
findColors = re.compile(r'(red)|(green)|(blue)')

maxVals = {'red':12,'green':13,'blue':14}

def testSample(samples):
    for sample in samples.split(';'):
        drawGroup = sample.split(',')
        for draw in drawGroup:
            nCubes = int(re.search(findNumbers,draw)[0])
            color = re.search(findColors,draw)[0]
            if nCubes > maxVals[color]:
                return False
    return True

tally = 0

for line in data:
    head,_,tail = line.partition(':')
    gameNum = re.search(findNumbers,head)[0]
    samples = tail.split(';')
    if testSample(tail):
        tally += int(gameNum)

print(tally)      
    
    



2593


2593 is correct.

Part II:
Now we need to find the max value for each color for each game:
* Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
* In game 1, the game could have been played with as few as 4 red, 2 green, and 6 blue cubes.

Find the product of the max values for each game and tally them.

I can reuse most of the code. But the test function will create a dictionary and update it with the max values for each sample. After looking at all the samples, it will return the product.



In [74]:
testData='''Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green'''
import re
findNumbers  = re.compile(r'\d+')
findColors = re.compile(r'(red)|(green)|(blue)')

def testSample(samples):
    maxVals = {'red':1,'green':1,'blue':1} # setting to 1 because if 0, it would make product 0
    for sample in samples.split(';'):
        drawGroup = sample.split(',')
        for draw in drawGroup:
            nCubes = int(re.search(findNumbers,draw)[0])
            color = re.search(findColors,draw)[0]
            if nCubes > maxVals[color]:
                maxVals[color] = nCubes
    
    return maxVals['red'] * maxVals['blue'] * maxVals['green']

tally = 0

for line in testData.split('\n'):
    head,_,tail = line.partition(':')
    gameNum = re.search(findNumbers,head)[0]
    samples = tail.split(';')
    tally += testSample(tail)
    print(tail,testSample(tail))

print(tally)

 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green 48
 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue 12
 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red 1560
 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red 630
 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green 36
2286


Ok, test data works. (that was easy)
On to the actual data...

In [76]:
with open ('2302input.txt') as f_in:
    data = f_in.read().split('\n')

import re
findNumbers  = re.compile(r'\d+')
findColors = re.compile(r'(red)|(green)|(blue)')

def testSample(samples):
    maxVals = {'red':1,'green':1,'blue':1} # setting to 1 because if 0, it would make product 0
    for sample in samples.split(';'):
        drawGroup = sample.split(',')
        for draw in drawGroup:
            nCubes = int(re.search(findNumbers,draw)[0])
            color = re.search(findColors,draw)[0]
            if nCubes > maxVals[color]:
                maxVals[color] = nCubes
    
    return maxVals['red'] * maxVals['blue'] * maxVals['green']

tally = 0

for line in data:
    head,_,tail = line.partition(':')
    gameNum = re.search(findNumbers,head)[0]
    samples = tail.split(';')
    tally += testSample(tail)

print(tally)

54699


54699 is correct for my set of data.